compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -716,11 +716,13 @@ pub enum ExistentialPredicate<'tcx> {
     AutoTrait(DefId),
 }
 
-impl<'tcx> ExistentialPredicate<'tcx> {
+pub type ExistentialPredicate<'tcx> = Binder<'tcx, WhereClause<'tcx>>;
```

Comment:
This isn't quite right. The difference between rustc and Chalk isn't solely the name. In Chalk, there's actually two layers of Binders. One that contains a single type that represents `Self`, and one that represents the vars on the predicates itself. See https://github.com/rust-lang/chalk/blob/802599ce2ca02f322c078f472e64c5cd3a208225/chalk-solve/src/rust_ir.rs#L233

Reply:
I guess the big thing missing here is the removal of `erase_self_ty`. And along with that, make the `Self` ty actually be a bound var.